File: Java/src/CatenableList.java
public interface CatenableList<T> extends Queue<T>
{
    @Override
    boolean isEmpty();

    CatenableList<T> cons(T value);

    @Override
    CatenableList<T> snoc(T value);

    CatenableList<T> concat(CatenableList<T> list);

    @Override
    T head();

    @Override
    CatenableList<T> tail();

    @Override
    void printWithoutConcat();

    @Override
    void toStringEfficient(StringBuilder sb);

}

File: Haskell/CatenableList.hs
module CatenableList (CatenableList(..)) where
    import Prelude hiding (head, tail, (++))
    import Queue

-- The value can be any type for example string or integer
class CatenableList catenableList where
    empty :: catenableList value
    isEmpty :: catenableList value -> Bool
    -- Takes a value and a catenable list as arguments
    -- Returns a new catenable list with the value as the first element
    cons :: value -> catenableList value -> catenableList value
    snoc :: catenableList value -> value -> catenableList value
    -- Takes two catenable lists as arguments
    -- Returns the concatenation as a new catenable list 
    (++) :: catenableList value -> catenableList value -> catenableList value
    head :: catenableList value -> value
    tail :: catenableList value -> catenableList value
        